docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+
+Good luck pal ;)
+
+## Proposal
+
+### Versions
+
+Each of BlockVersion, AppVersion, P2PVersion is a monotonically increasing int64.
+
+An alternative might be to try to use Amino - ie. include the version number in the registration name for some type,
+for instance the BlockVersion in the header. However this would require the registered-name of the header to change even
+if it wasn't the header type itself that changed, which could be confusing. Thus, versions here will be integers in Amino encoded
+structus. That said, Amino's flexibility (ie. both in registration of new types and proto3 extensibility in structs) will make
+actual upgrades smoother.
+
+#### BlockVersion
+
+BlockVersion is the first field in the block Header. Since we have settled on a proto3 header,
+the ability to read the BlockVersion out of the serialized header is unanimous.
+
+BlockVersion is included in the NodeInfo and we only connect to peers with the same or higher BlockVersion.
+
+In the event that we need to upgrade the core protocols and don't want to dump the state,
+we can condition execution on the block version and use different handlers / message types as needed.
+In such a case we may need to support connecting to old versions as well.
+
+This isn't nice, but should happen quite infrequently and ideally only for extreme emergency.
+
+Note Ethereum has not had to make an upgrade like this (everything has been at state machine level, AFAIK).
+
+#### P2PVersion
+
+P2PVersion is not included in the block Header.
+
+P2PVersion is the first field in the NodeInfo. NodeInfo is also proto3 so this is easy to read out.
+
+For each P2PVersion, we keep a list of previous versions that this one is compatible with so we can connect to
+peers of any of those versions or higher.
+
+Note we need the peer/reactor protocols to take the versions of peers into account when sending messages:
+    - don't send messages they don't understand
+    - don't send messages they don't expect
+Doing this will be specific to the upgrades being made.
+
+Note we also include the list of reactor channels in the NodeInfo and already don't send messages for channels the peer doesn't understand.
+If upgrades always use new channels, this simplifies the development cost of backwards compatibility.
+
+Note NodeInfo is only exchanged after the authenticated encryption handshake to ensure that it's private.
+Doing any version exchange before encrypting could be considered information leakage, though I'm not sure
+how much that matters compared to being able to upgrade the protocol.
+
+XXX: if needed, can we change the meaning of the first byte of the first message to encode a handshake version?
+this is the first byte of a 32-byte ed25519 pubkey.
+
+#### AppVersion
+
+AppVersion is also included in the block Header and the NodeInfo.
+
+We only maintain connections to peers with the correct AppVersion for the height they're at.
```

Comment:
I don't understand this. Can a peer update app version but stay with the same Tendermint version? Why we shouldn't connect to that peer?

Reply:
Correct. Because an updated AppVersions means they will execute transactions differently so we will not be able to agree on the results of a block.

Reply:
Very good point, thanks.